### ISSUE TITLE
Create helper script to run visualization with parameters from riboviz run

### ISSUE THREAD

**Issue by FlicAnderson** [riboviz/riboviz#275]
From discussion & [suggestion](https://github.com/riboviz/riboviz/issues/239#issuecomment-766787980) in #239: 

Create a Nextflow process which produces a 'helper' script which is the run command for the interactive visualization system #239 

This script would: 
* run to produce the #239 visualization system 
* contain the relevant parameters & filepaths for the dataset run through riboviz, as fed in by the nextflow process.

:question: should the script be output into the dataset folder (or more specifically `dir_out`) by default? (and Should we have an option to specify the output location as part of the script if you want to put it somewhere non-default?)

**Comment by FlicAnderson** [comment]
No progress made on this yet.

**Comment by FlicAnderson** [comment]
This ticket is likely next step after merging of code for riboviz/#193 (static html output per sample). 

At that point I can swap focus to creating this, probably building on the handy groovy-code 'script-building' technique @mikej888 used in `prep_riboviz.nf` for the `staticHTML` process ([described here](https://github.com/riboviz/riboviz/issues/193#issuecomment-789919728)).  

The chunk of code within `staticHTML` handles conditional files and hands them into `rmarkdown/AnalysisOutputs.Rmd` from `staticHTML` process.  I think this is probably pretty similar to what we'd need to hand the correct file information into the visualization code (riboviz/#239).

**Comment by FlicAnderson** [comment]
@john-s-f and @p6t do you have your visualisation code pushed up to riboviz just now? It would be great to know what command you're currently running to generate your visualisation? 

I could then use this as the basis of work on this ticket (#275) to create the helper script.  Even if it's not 100% final, it would be really useful to have something to work with to start with.  Presumably in the same way we've got the .tsv files handed into the .Rmd file you'd need at least some of that information passed over to your visualisation run command?

If you could add a bit of information on this ticket about what you need passed out from nextflow I can start working on this asap. Thanks!

**Comment by john-s-f** [comment]
Currently I'm trying to make it `Rscript --vanilla make_shiny_server.R path/to/yaml/config.yaml`, so all you have to do is provide the path to the YAML and it works out. I'll try and push later today when I fix some remaining broken things.

**Comment by mikej888** [comment]
> :question: what type of file should the helper script output?  `.txt`? `.sh`?   If we want a shell script, do we need to do anything special (ie somehow make it executable)? @john-s-f @mikej888 - any strong feelings about this? Perhaps a script might involve less 'user barriers' and be slightly nicer? Or would a .txt they can copy and paste from be sufficent? 

It'd be more usable if it was a shell script rather than something users copy-paste-execute.

The shell script doesn't have to be executable as the user can run, for example, `bash interactive_viz.sh` regardless.

> I'll need to create a new config file parameter: `create_helper_script` or some similar name. 
:question: do we want this to be TRUE by default, so that this script is always created? Or do we want this to be false? 

If it's only a line or two then I'd agree with @john-s-f and just create it regardless.

> Alternatively, the visualization system only requires the path to the yaml config file, so if it's possible that we can have nextflow modify the Rscript to insert the path into it, then the user only need run the Rscript. Perhaps we can make a spot in our Rscript that is like REPLACE_ME_WITH_YAML and do a simple shell command like sed -i 's/REPLACE_ME_WITH_YAML/$path_to_yaml_from_nextflow/' shiny_server.R

The desirability of this might depend on how big shiny_server.R is. It won't be edited, but copied, edited, then output into `dir_output`. If it's a couple of lines then fine. However, if it's a couple of hundred then having it take the path as a parameter and outputting a very short helper script is preferable. It makes testing shiny_server.R without editing it easier, and also it does not take Ribdown down the path of becoming a code generator!

If the visualisation system needs the YAML file created by Nextflow as an input, then that too needs to be published by the Nextflow task into `dir_output` (using the YAML file _given_ to Nextflow as-is will not be possible as, at some point will support environment variable tokens which will need to be handled by any code that reads such YAML files).

> Reminder to self: I'll have to go through and add this parameter into all the required places (further up in prep_riboviz.nf, into the config documentation, etc). 

Reminder, https://github.com/riboviz/riboviz/blob/develop/docs/developer/adding-config.md

**Comment by john-s-f** [comment]
@FlicAnderson The script is in a good shape, it runs quite fine on the vignette datasets. The only thing it's missing is some work with the `here` package, which will ensure all of the paths always work out correctly. Currently it has the option to provide a path to the yaml at the command line, but I think it's the case that these need to be run from in Rstudio so that might be useless. 

Some things for @p6t and I to look at.
- [ ] Use `here` to make sure paths always work out, currently it's linked to a yaml in my directory structure to ensure all the plots work but this cannot remain.
- [ ] Last plot to add, genewise reads across a gene from the H5. This needs to be smart about how it gets data from the H5 as to not slow things down. 
- [ ] Look for possible speed improvements, for example, wait a few seconds for the user to be done checking boxes until it starts processing. There is also a submit button you can add. It's simpler if the user doesn't have to push a button but that could be a last resort

**Comment by FlicAnderson** [comment]
@john-s-f I think I missed replying to your comment above sooner, sorry.  

> Alternatively, the visualization system only requires the path to the yaml config file, so if it's possible that we can have nextflow modify the Rscript to insert the path into it

This is what the helper script is aiming to do. It will create a (I think we've decided) shell script containing the correct path to the config file & shiny script name, so you could just run the script and the correct config file path should be in place.  

Is this what you were expecting?

**Comment by john-s-f** [comment]
I think that is fine, I've pushed the most recent changes and everything seems functional when I explicitly give it a path to the yaml on line 20. Currently, there is code at the top of script which I believe allows it to take an argument from the command line. Presumably, the helper script ends up as `Rscript --vanilla shiny_final.R config.yaml`. I think this would work if the code at the top truly allows it to accept an argument from the command line.

The only other issue is the function that gets the sample names from the yaml. Specifically, lines 29 and 39 have hard coded path values in them that definitely won't work if this gets moved out of my directory. I assume the `here` package solves this, but I'm not entirely sure how we'll use `here` in conjunction with the helper script to solve this.

**Comment by FlicAnderson** [comment]
As of commit f64c54b  in branch `viz-helper-275`, I've edited the shell section so it creates a new file `interactive_viz.sh` and fills it with `Rscript -e`, the filepath for `rscripts/shiny_final.R` (:eyes: note the name needs to match the eventual name of the file currently named `shiny_final.R` from #239), and after another space, includes the path to the 'copy' config file.  

This last part needs fixing, because currently when I run prep_riboviz.nf, it works, and generates `interactive_viz.sh`, but these are the contents: 
```
Rscript -e '/home/fanders6/riboviz/riboviz/rscripts/shiny_final.R' '/home/fanders6/riboviz/riboviz/work/21/7acba2be73d831e352d7ab73a25134/config.yaml'
``` 

The working directory path for the config file isn't the best.  I think this must be what @mikej888 mentioned [earlier](https://github.com/riboviz/riboviz/issues/275#issuecomment-823121678):
> If the visualisation system needs the YAML file created by Nextflow as an input, then that too needs to be published by the Nextflow task into dir_output (using the YAML file given to Nextflow as-is will not be possible as, at some point will support environment variable tokens which will need to be handled by any code that reads such YAML files).  

I'm not sure how to handle this.  I mean, the working directory specified does contain the copy config file and the interactive viz script, so it probably suffices for testing the shiny visualisation at least (though due to errors I haven't managed to try this yet), but for users it'd probably help if it referred to the main config file somehow? 

Question (@mikej888 especially) - does this interactive_viz.sh output need a shebang line in it or anything fancy or extra?

**Comment by mikej888** [comment]
In `helperViz`, the line:
```
publishDir "${params.dir_out}", mode: 'copy', overwrite: true
```
should already copy `interactive_viz.sh` to `${params.dir_output}` as it is declared in the task's `output` block. Declaring the YAML file as an output would publish the YAML to the output directory likewise, so the task becomes something like:

```
output:
  file "interactive_viz.sh" into interactive_viz_sh
  file "config.yaml" into config_yaml
shell:
  script = "Rscript -e "
  script += "'${workflow.projectDir}/rscripts/shiny_final.R' "
  script += "'${params.dir_out}/config.yaml'"
  """
  echo "${config_file_yaml}" > "config.yaml"
  echo "${script}" > "interactive_viz.sh"
  """
```
This would also resolve the issue that the following is too restrictive,
```
script += "'\$PWD/${config_file_yaml}'"
```
as one cannot guarantee that the workflow will be run from within the same directory as the configuration file.

On a related note, the line:
```
config_yaml = new Yaml().dump(params)
```
will need a rewrite once Add support for use of environment variable names in YAML configuration file paths #274 is merged since file-related parameters will be preprocessed to replace environment variables so, once that branch is merged, `params` will not be used to access any file-related parameters (https://github.com/riboviz/riboviz/issues/274#issuecomment-802054848). It would help when merging that branch if, instead of dumping the whole YAML file into the YAML variable, you create a variable with only those parameters needed by the `shiny_final.R` (this also makes it clearer to those reading the workflow what parameters of the original configuration are used by `shiny_final.R` without them having to look at `shiny_final.R` itself). An example of this is done when preparing the YAML input for `countReads`:

```
Map ribosome_fqs = [:]
ribosome_fqs.fq_files = params.fq_files
ribosome_fqs.multiplex_fq_files = params.multiplex_fq_files
ribosome_fqs_yaml = new Yaml().dump(ribosome_fqs)
```

which is used as an input to the `countReads` task. So for `shiny_final.R` you may want something like:
```
Map analysis_outputs_inputs = [:]
if (is_asite_disp_length_file) {
    analysis_outputs_inputs.asite_disp_length_file = params.asite_disp_length_file
}
if (is_codon_positions_file) {
    analysis_outputs_inputs.codon_positions_file = params.codon_positions_file
}
if (is_features_file) {
    analysis_outputs_inputs.features_file = params.features_file
}
if (is_t_rna_file) {
    analysis_outputs_inputs.t_rna_file = params.t_rna_file
}
analysis_outputs_inputs_yaml = new Yaml().dump(analysis_outputs_inputs_yaml)
```
(adding whatever other parameters are needed), and then use that instead of `config_yaml` as an input to `helperViz`.

> Question (@mikej888 especially) - does this interactive_viz.sh output need a shebang line in it or anything fancy or extra?

A shebang line is optional as the user can do:
```console
$ bash interactive_viz.sh
```
But if it's straightforward to add then add it. Then, see if the Nextflow task can do:
```
chmod +x interactive_viz.sh
```
after the `echo`, so the user could then run:
```console
$ ./interactive_viz.sh
```

**Comment by FlicAnderson** [comment]
I've (kind of) made the changes suggested by @mikej888 [here](https://github.com/riboviz/riboviz/issues/275#issuecomment-843915389).  

I tried making the changes as-is initially, however this led to issues with re-use of the config-yaml copy channel, so in the end, I created a new process `createHelperConfigFile` which creates a new duplicate yaml specifically for use by `helperViz`, which is filled with the parameters correctly.  I added the `publishDir "${params.dir_out}", mode: 'copy', overwrite: true` to the `createHelperConfigFile` process instead of the `helperViz` one, as this seemed to correctly generate the file in the right place, and made sense (to me at least). 

Riboviz runs all processes without problems as of commit ec6338e9, here's details from running on the plain vignette: 
```
(riboviz) fanders6@ceres:~/riboviz/riboviz$ nextflow run prep_riboviz.nf -params-file vignette/vignette_config.yaml -ansi-log false
N E X T F L O W  ~  version 20.07.1
Launching `prep_riboviz.nf` [jovial_franklin] - revision: 26391e83a0
No such sample file (NotHere): example_missing_file.fastq.gz
[fd/e6cbaf] Submitted process > cutAdapters (WTnone)
[0a/8e8000] Submitted process > buildIndicesrRNA (yeast_rRNA)
[f6/30294b] Submitted process > buildIndicesORF (YAL_CDS_w_250)
[f6/80b1d3] Submitted process > cutAdapters (WT3AT)
[20/cea80c] Submitted process > createConfigFile
[e2/bebf54] Submitted process > createHelperConfigFile
[5a/f20690] Submitted process > helperViz
[0e/37e17a] Submitted process > hisat2rRNA (WTnone)
[f8/11169e] Submitted process > hisat2rRNA (WT3AT)
[0c/427b0f] Submitted process > hisat2ORF (WTnone)
[f8/94ad31] Submitted process > trim5pMismatches (WTnone)
[08/bc3aaa] Submitted process > samViewSort (WTnone)
[10/cd1a3c] Submitted process > outputBams (WTnone)
[1b/bf7d22] Submitted process > makeBedgraphs (WTnone)
[49/29e443] Submitted process > bamToH5 (WTnone)
[39/de1263] Submitted process > hisat2ORF (WT3AT)
[be/c8f549] Submitted process > trim5pMismatches (WT3AT)
[fa/dd35c7] Submitted process > samViewSort (WT3AT)
[25/cbf9fa] Submitted process > outputBams (WT3AT)
[43/60b668] Submitted process > makeBedgraphs (WT3AT)
[6d/7cb6e3] Submitted process > bamToH5 (WT3AT)
[60/1fd421] Submitted process > generateStatsFigs (WTnone)
[4f/3b6f11] Submitted process > generateStatsFigs (WT3AT)
Finished processing sample: WTnone
[28/30e847] Submitted process > renameTpms (WTnone)
[af/d71b85] Submitted process > staticHTML (WTnone)
Finished processing sample: WT3AT
[52/492cd0] Submitted process > renameTpms (WT3AT)
[5d/a7a5f0] Submitted process > staticHTML (WT3AT)
[30/a21aec] Submitted process > collateTpms (WTnone, WT3AT)
[a3/92e7fb] Submitted process > countReads
Finished visualising sample: WTnone
Finished visualising sample: WT3AT
Workflow finished! (OK)
``` 

This is the content of the output directory: 
```
(riboviz) fanders6@ceres:~/riboviz/riboviz$ ls vignette/output/
helper_config.yaml  interactive_viz.sh  read_counts.tsv  TPMs_collated.tsv  WT3AT  WTnone
```
:heavy_plus_sign:  New files generated in the output directory by code for current issue (#275) are: `helper_config.yaml` and `interactive_viz.sh`.  
:thinking: These new items will need tests for them at some point before this is completed.

This is the current output of the helper script `output/interactive_viz.sh`: 
```
Rscript -e '/home/fanders6/riboviz/riboviz/rscripts/run_shiny_server.R' 'vignette/output/helper_config.yaml'
```  
(:exclamation:  Note, I've changed the name of the R viz script from `shiny_final.R` to `run_shiny_server.R` for now - this means it's currently inconsistent with #239  script naming, but that code isn't currently in branch `viz-helper-275` anyway, so this needs checking/fixing on integration!)

The only thing which might be relevant to #239 is how the `fq_files` parameter is formatted inside the `helper_config.yaml`: 
`fq_files: {WTnone: SRR1042855_s1mi.fastq.gz, WT3AT: SRR1042864_s1mi.fastq.gz, NotHere: example_missing_file.fastq.gz}`  

I'm not sure if this would affect how `yaml::read_yaml()` function works or not? '{}' might be valid yaml syntax, I'm not certain and havent' checked yet.

**Comment by mikej888** [comment]
Rather than a new `createHelperConfigFile` process you can copy the output, `config_file_yaml` from `createConfigFile` process, using `into` e.g. something like:

```
// Split channel for use in multiple downstream processes.
config_yaml.into {
    static_html_config_yaml; helper_viz_config_yaml
}

process staticHTML {
  ...
  input:
    file config_file_yaml from static_html_config_yaml
  ...
}

process helperViz {
  ...
  input:
    file helper_config_file_yaml from helper_viz_config_yaml
  ...
```

In YAML, `{...}` is a more succinct, but less human-readable, way of writing a dictionary, so:
```
fq_files: {WTnone: SRR1042855_s1mi.fastq.gz, WT3AT: SRR1042864_s1mi.fastq.gz, NotHere: example_missing_file.fastq.gz}
```
is the same as 
```
fq_files:
  WTnone: SRR1042855_s1mi.fastq.gz
  WT3AT: SRR1042864_s1mi.fastq.gz
  NotHere: example_missing_file.fastq.gz
```
For example:
```console
$ cat example.yaml
fq_files: {WTnone: SRR1042855_s1mi.fastq.gz, WT3AT: SRR1042864_s1mi.fastq.gz, NotHere: example_missing_file.fastq.gz}
$ cat alt-example.yaml
fq_files:
  WTnone: SRR1042855_s1mi.fastq.gz
  WT3AT: SRR1042864_s1mi.fastq.gz
  NotHere: example_missing_file.fastq.gz
```
```python
>>> import yaml
>>> with open("example.yaml", "r") as f:
...     example = yaml.load(f, yaml.SafeLoader)
... 
>>> with open("alt-example.yaml", "r") as f:
...     alt_example = yaml.load(f, yaml.SafeLoader)
... 
>>> example
{'fq_files': {'WTnone': 'SRR1042855_s1mi.fastq.gz', 'WT3AT': 'SRR1042864_s1mi.fastq.gz', 'NotHere': 'example_missing_file.fastq.gz'}}
>>> example == alt_example
True
```

**Comment by FlicAnderson** [comment]
> Rather than a new `createHelperConfigFile` process you can copy the output, `config_file_yaml` from `createConfigFile` process, using `into` e.g. something like:
> 
> ```
> // Split channel for use in multiple downstream processes.
> config_yaml.into {
>     static_html_config_yaml; helper_viz_config_yaml
> }
> 
> process staticHTML {
>   ...
>   input:
>     file config_file_yaml from static_html_config_yaml
>   ...
> }
> 
> process helperViz {
>   ...
>   input:
>     file helper_config_file_yaml from helper_viz_config_yaml
>   ...
> ```
Thanks @mikej888, I knew there was a smarter way of handling the channel-shenanigans, but my mind was in talk-prep mode!  I'll have a bash at this when I have a moment. 

I think this (V) has probably come up before and I'd forgotten - thanks!  
> In YAML, `{...}` is a more succinct, but less human-readable, way of writing a dictionary, so:
> 
> ```
> fq_files: {WTnone: SRR1042855_s1mi.fastq.gz, WT3AT: SRR1042864_s1mi.fastq.gz, NotHere: example_missing_file.fastq.gz}
> ```
> 
> is the same as
> 
> ```
> fq_files:
>   WTnone: SRR1042855_s1mi.fastq.gz
>   WT3AT: SRR1042864_s1mi.fastq.gz
>   NotHere: example_missing_file.fastq.gz
> ```
> 
> For example:
> 
> ```
> $ cat example.yaml
> fq_files: {WTnone: SRR1042855_s1mi.fastq.gz, WT3AT: SRR1042864_s1mi.fastq.gz, NotHere: example_missing_file.fastq.gz}
> $ cat alt-example.yaml
> fq_files:
>   WTnone: SRR1042855_s1mi.fastq.gz
>   WT3AT: SRR1042864_s1mi.fastq.gz
>   NotHere: example_missing_file.fastq.gz
> ```
> 
> ```python
> >>> import yaml
> >>> with open("example.yaml", "r") as f:
> ...     example = yaml.load(f, yaml.SafeLoader)
> ... 
> >>> with open("alt-example.yaml", "r") as f:
> ...     alt_example = yaml.load(f, yaml.SafeLoader)
> ... 
> >>> example
> {'fq_files': {'WTnone': 'SRR1042855_s1mi.fastq.gz', 'WT3AT': 'SRR1042864_s1mi.fastq.gz', 'NotHere': 'example_missing_file.fastq.gz'}}
> >>> example == alt_example
> True
> ```

**Comment by mikej888** [comment]
@FlicAnderson When merging branch `yaml-environment-vars-274` into `develop` (as part of closing #274) I got some environment variable-related test failures arising from the use of:
```
config_yaml = new Yaml().dump(params)
```
in `prep_riboviz.nf` as `config_yaml` ended up with parameter values with un-replaced tokens. To fix the workflow to get the tests running I've refactored the above to create a YAML file with only the parameters required by `staticHTML`:
```
Map viz_params = [:]
if (is_asite_disp_length_file) {
    viz_params.asite_disp_length_file = asite_disp_length_file
}
if (is_codon_positions_file) {
    viz_params.codon_positions_file = codon_positions_file
}
if (is_features_file) {
    viz_params.features_file = features_file
}
if (is_t_rna_file) {
    viz_params.t_rna_file = t_rna_file
}
viz_params_yaml = new Yaml().dump(viz_params)
```
`createConfigFile` has been renamed and updated to:
```
process createVizParamsConfigFile {
    input:
      val viz_params_yaml from viz_params_yaml
     output:
      file "config.yaml" into viz_params_config_file_yaml
    shell:
      """
      echo "${viz_params_yaml}" > "config.yaml"
      """
}
```
`staticHTML` has been updated to:
```
process staticHTML {
    ...
    publishDir "${dir_out}/${sample_id}", \
    input:
      file viz_params_config_file_yaml from viz_params_config_file_yaml
      ...
    shell:
      ...
      script += "yamlfile='\$PWD/${viz_params_config_file_yaml}', "
      ...
    }
```
Note also the use of `${dir_out}` and not `${params.dir_out}` as `dir_out` is `${params.dir_out}` with tokens replaced.

This is the pattern I'd suggested above, as a consequence of the requirements described in https://github.com/riboviz/riboviz/issues/274#issuecomment-802054848.

**Comment by FlicAnderson** [comment]
> > Rather than a new `createHelperConfigFile` process you can copy the output, `config_file_yaml` from `createConfigFile` process, using `into` e.g. something like:
> > ```
> > // Split channel for use in multiple downstream processes.
> > config_yaml.into {
> >     static_html_config_yaml; helper_viz_config_yaml
> > }
> > 
> > process staticHTML {
> >   ...
> >   input:
> >     file config_file_yaml from static_html_config_yaml
> >   ...
> > }
> > 
> > process helperViz {
> >   ...
> >   input:
> >     file helper_config_file_yaml from helper_viz_config_yaml
> >   ...
> > ```
> 
> Thanks @mikej888, I knew there was a smarter way of handling the channel-shenanigans, but my mind was in talk-prep mode! I'll have a bash at this when I have a moment.
> 
> I think this (V) has probably come up before and I'd forgotten - thanks!
> 
> > In YAML, `{...}` is a more succinct, but less human-readable, way of writing a dictionary, so:
> > ```
> > fq_files: {WTnone: SRR1042855_s1mi.fastq.gz, WT3AT: SRR1042864_s1mi.fastq.gz, NotHere: example_missing_file.fastq.gz}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > is the same as
> > ```
> > fq_files:
> >   WTnone: SRR1042855_s1mi.fastq.gz
> >   WT3AT: SRR1042864_s1mi.fastq.gz
> >   NotHere: example_missing_file.fastq.gz
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > For example:
> > ```
> > $ cat example.yaml
> > fq_files: {WTnone: SRR1042855_s1mi.fastq.gz, WT3AT: SRR1042864_s1mi.fastq.gz, NotHere: example_missing_file.fastq.gz}
> > $ cat alt-example.yaml
> > fq_files:
> >   WTnone: SRR1042855_s1mi.fastq.gz
> >   WT3AT: SRR1042864_s1mi.fastq.gz
> >   NotHere: example_missing_file.fastq.gz
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```python
> > >>> import yaml
> > >>> with open("example.yaml", "r") as f:
> > ...     example = yaml.load(f, yaml.SafeLoader)
> > ... 
> > >>> with open("alt-example.yaml", "r") as f:
> > ...     alt_example = yaml.load(f, yaml.SafeLoader)
> > ... 
> > >>> example
> > {'fq_files': {'WTnone': 'SRR1042855_s1mi.fastq.gz', 'WT3AT': 'SRR1042864_s1mi.fastq.gz', 'NotHere': 'example_missing_file.fastq.gz'}}
> > >>> example == alt_example
> > True
> > ```

Just as an update, I haven't had any time to work on this and probably won't get a chance before I'm off on my annual leave.  I suspect Mike's fixes in the comment above probably supercede what I'd planned to do?  Is that right @mikej888 ?  Or are these only related to the environmental variables branch code?  (Apologies, I can't remember whether that got merged yet)
 
(This comment overall is particularly aimed at @p6t and @john-s-f to let you know where I left off.)

**Comment by mikej888** [comment]
@FlicAnderson the YAML environment branch was merged into `develop` last week so you can get the updates I made. I think they supercede some of what you'd planned to do but if you could check that would be good :-)

**Comment by FlicAnderson** [comment]
Current status: 

I attempted to merge in the environmental variable changes from `develop` locally, but wasn't able to figure out how to integrate @mikej888's updates from merging the #274 work with the code I currently had and get it working correctly, **so I have NOT pushed this merge of `develop` into `viz-helper-275`**.  I think it might be better for whoever is going to take this issue up in the next week or two to start off with the code in `viz-helper-275` and then merge develop in theirselves and start from there, rather than starting from a post-merge state that doesn't work. 

The key aims for completing this issue will be: 

* **ensure the non-original (copy) config yaml file is duplicated correctly** (need to take account of @mikej888's advice on how to duplicate the config file) for use by both staticHTML and helperViz processes [see suggestion on how to do this using .into method:](https://github.com/riboviz/riboviz/issues/275#issuecomment-844044237) )
* ensure helper viz script generated by Nextflow helperViz process **contains the correct (current) R script name for the shiny code** (#239)   
* ensure helper script `interactive_viz.sh` generated by Nextflow **helperViz process accurately outputs the path and correct filename for a non-original config yaml file** (non-original, ie a copy, because this is a long-term issue that Nextflow can't take the main configuration yaml file as an input within the main workflow!) 
* **ensure that the non-original config yaml file contains the correct yaml parameters and is output to the correct location** (path matches the path in `interactive_viz.sh` (created by helperViz nextflow process).

**Comment by mikej888** [comment]
#356 has now been merged into `develop`. This fixes a bug whereby the YAML file output by `staticHTML` contained Groovy file class names and not file paths.

**Comment by FlicAnderson** [comment]
Based on discussion at developer meeting (28 Jul), we don't need to generate a 'helper script' containing the 1-line of code to run the interactive viz, but we DO still need a yaml file generated from Nextflow which then would be taken and used as the input for the command to run the interactive visualization #239.  This generated yaml file would contain any path location Nextflow had processed from the environment variable tokens in the original yaml.

Need to make sure that the nextflow process hands in the correct information from the original config file to the newly generated config output file so the interactive code can pick it up.

**Comment by mikej888** [comment]
An example of how to do this is already in `prep_riboviz.nf` where you create the YAML file for `staticHTML`:
```
Map viz_params = [:]
if (is_asite_disp_length_file) {
    viz_params.asite_disp_length_file = asite_disp_length_file.toString()
}
if (is_codon_positions_file) {
    viz_params.codon_positions_file = codon_positions_file.toString()
}
if (is_features_file) {
    viz_params.features_file = features_file.toString()
}
if (is_t_rna_file) {
    viz_params.t_rna_file = t_rna_file.toString()
}
viz_params_yaml = new Yaml().dump(viz_params)

process createVizParamsConfigFile {
    input:
      val viz_params_yaml from viz_params_yaml
     output:
      file "config.yaml" into viz_params_config_file_yaml
    shell:
      """
      echo "${viz_params_yaml}" > "config.yaml"
      """
}
```
You could create a similar `Map` and `process` for the parameters required for the Shiny app to the `Map` and also ensuring that the `process` then publishes the YAML to the output directory by adding the following to the `process`:
```
    publishDir "${dir_out}", mode: 'copy', overwrite: true
```

**Comment by FlicAnderson** [comment]
In branch `viz-helper-275` I've edited `prep_riboviz.nf` to include a new process (`createInteractiveVizParamsConfigFile`) which generates a new config file specifically for use by #239's  `run_shiny_server.R`.  It contains parameters: `dir_in`, `fq_files` and `sample_sheet`.  

@mikej888 I see that `dir_in` is the only parameter of those three that could possibly use an environment token, as the other two are relative to `dir_in`.  Can you confirm whether I understand correctly that [this line of code](https://github.com/riboviz/riboviz/blob/viz-helper-275/prep_riboviz.nf#L439) resolves the environment tokens for all processes/mapping further down the nextflow script so that my new process `createInteractiveVizParamsConfigFile` doesn't need to do anything special to include and use the environmental token path information?

**Comment by mikej888** [comment]
@FlicAnderson Yes, that's right, so long as you refer to `dir_in` (and not `params.dir_in`). The 11 parameters that allow environment tokens (https://github.com/riboviz/riboviz/blob/develop/docs/user/prep-riboviz-config.md#configuring-file-paths-and-directories) are all preprocessed at the top of the Nextflow file when validating the configuration.

**Comment by FlicAnderson** [comment]
> @FlicAnderson Yes, that's right, so long as you refer to `dir_in` (and not `params.dir_in`). The 11 parameters that allow environment tokens (https://github.com/riboviz/riboviz/blob/develop/docs/user/prep-riboviz-config.md#configuring-file-paths-and-directories) are all preprocessed at the top of the Nextflow file when validating the configuration.

Aha! Thanks, I had used `params.dir_in`, commit e58f8d33dd in branch `viz-helper-275` fixes that and uses just `dir_in` now.

**Comment by FlicAnderson** [comment]
Added [more parameters used by run_shiny_server.R](https://github.com/riboviz/riboviz/issues/239#issuecomment-892750921) to the `interactive_viz_params` Map and `createInteractiveVizParamsConfigFile` process.  

Example contents of the new interactive viz config file: 

After running simdata_multiplex vignette: `nextflow run prep_riboviz.nf -params-file vignette/simdata_multiplex_config.yaml -ansi-log false`, the new config file is generated in the `dir_out` location (in this case, vignette/simdata_multiplex_output/) and contains: 

```
{dir_in: data/simdata/, 
dir_out: vignette/simdata_multiplex_output, 
dataset: simdata_multiplex,
fq_files: null, 
sample_sheet: multiplex_barcodes.tsv, 
features_file: /home/fanders6/riboviz/riboviz/data/yeast_features.tsv}
```

**Comment by FlicAnderson** [comment]
@mikej888 suggested it's worth checking whether this code generates absolute paths or relative paths for the `interactive_viz_config.yaml` file.  

Also, this code needs testing before sign-off (can currently be tested from branch `viz-helper-275` on its own, or within `visualization-239-FlicAnderson` to test the interactive visualization using this newly generated `interactive_viz_config.yaml` file).

**Comment by mikej888** [comment]
@FlicAnderson it looks like the paths will be absolute in the output YAML. I had an example for `vigentte_config.yaml` as output by your `createVizParamsConfigFile` task that creates the YAML for the `staticHTML` task:

```console 
$ cat work/5f/4cae53b3be55555600292f0691aaa6/config.yaml 
{asite_disp_length_file: /home/ubuntu/riboviz/data/yeast_standard_asite_disp_length.txt,
  codon_positions_file: /home/ubuntu/riboviz/data/yeast_codon_pos_i200.RData, features_file: /home/ubuntu/riboviz/data/yeast_features.tsv,
  t_rna_file: /home/ubuntu/riboviz/data/yeast_tRNAs.tsv}
```

**Comment by FlicAnderson** [comment]
@mikej888 hmmm, my example [above](https://github.com/riboviz/riboviz/issues/275#issuecomment-892772163) seems to have just relative paths for some of the parameters (e.g. `dir_in`), from the `createInteractiveVizParamsConfigFile` process.  I'm not 100% on how best to handle these paths in Nextflow, so I'd welcome any suggestions for how best to operate this. 

I ran the standard vignette_config,  and got this content for `vignette/output/interactive_viz_config.yaml`: 
```
dir_in: vignette/input
dir_out: vignette/output
dataset: vignette
fq_files: {WTnone: SRR1042855_s1mi.fastq.gz, WT3AT: SRR1042864_s1mi.fastq.gz, NotHere: example_missing_file.fastq.gz}
sample_sheet: null
features_file: /home/fanders6/riboviz/riboviz/data/yeast_features.tsv
```  
It seems like `features_file` is the only absolute path, is that to do with how `features_file` is handled since it's a conditional file?   

Would absolute paths be preferred for this `interactive_viz_config.yaml` file?  Is there a  groovy equivalent of Python's `file(dir_in).toAbsolutePath()`?  What would be the "most riboviz" method of handling the paths?

**Comment by FlicAnderson** [comment]
I think the absolute-paths-ness for files of this code should probably be documented, but since this code has now been merged into `visualization-239-john-s-f`, I think it might make sense to close this issue, and edit the documentation within another issue, or as part of a pull request for #239, since the work will continue in that branch.  

I'm advising closing this issue since `viz-helper-275` branch has been has been merged into `visualization-239-john-s-f`, will request feedback at dev meeting.

**Comment by ewallace** [comment]
When this code from @FlicAnderson is merged into @john-s-f's code, we can put a pull request into develop and close the issue. Within the next week?

visualization-239-FlicAnderson > visualization-239-john-s-f > develop

@john-s-f will create a separate issue ticket on the question of making the interactive viz work when output is downloaded to a different computer.

**Comment by FlicAnderson** [comment]
This code has been merged into #239 branches `visualization-239-FlicAnderson` and thence to `visualization-239-john-s-f`.  

Pull request #417 will merge this code to `develop` branch. Closing this issue.
